pkg/operation/botanist/component/kubestatemetrics/kube_state_metrics.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubestatemetrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ManagedResourceName is the name of the managed resource for seeds.
+	ManagedResourceName      = "kube-state-metrics"
+	shootManagedResourceName = "shoot-core-" + ManagedResourceName
+)
+
+// New creates a new instance of DeployWaiter for the kube-state-metrics.
+func New(
+	client client.Client,
+	namespace string,
+	secretsManager secretsmanager.Interface,
+	values Values,
+) component.DeployWaiter {
+	k := &kubeStateMetrics{
+		client:         client,
+		secretsManager: secretsManager,
+		namespace:      namespace,
+		values:         values,
+	}
+
+	if values.ClusterType == component.ClusterTypeSeed {
+		k.registry = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+	} else {
+		k.registry = managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
+	}
```

Comment:
Nit: I see that this is used in multiple places, probably you could also extract it in a method in the `component` package. E.g.
```
k.registry = component.NewManagedResourceRegistry(values.ClusterType)
```

Reply:
Not sure, TBH, I would rather keep it like this (there is not much gain with introducing yet another helper function)